preshed/bloom.pyx

```diff
@@ -54,40 +70,124 @@ cdef class BloomFilter:
+
+    bloom.hcount = hcount
+    bloom.length = length
+    # Technically overflow is possible here, but in valid data nothing will be
+    # lost.
+    bloom.seed = <uint32_t>seed
+
+    cdef key_t buflen = length // KEY_BITS
+    if length % KEY_BITS > 0:
+        buflen += 1
+    assert buflen > 0, "Tried to allocate an empty buffer"
+
+    contents = data[STRUCT_SIZE:]
+    bloom.bitfield = <key_t*>mem.alloc(buflen, sizeof(key_t))
+    memcpy(bloom.bitfield, <char*>contents, sizeof(key_t) * buflen)
+
+
+cdef void bloom_from_bytes_legacy(Pool mem, BloomStruct* bloom, bytes data):
+    # Older versions of this library used the array module with type L for
+    # serialization. Types in array guarantee a minimum size, not an actual
+    # size, and it turns out L is 8 bytes on Linux and most platforms, but 4 on
+    # Windows. 
+
+    # As a separate issue, due to bits/bytes confusion, each container in the
+    # serialized data has only one byte actually used.
+
+    # The code in this function reads in data in the old format and converts it
+    # to the current format losslessly. It also packs the significant bytes into
+    # contiguous memory.
+
+    # on non-Windows platforms
+    unit = "Q"
+    unit_size = LEGACY_UNIT_SIZE # size of container in bytes
+    cdef uint32_t offset = LEGACY_OFFSET
+
+    hcount, length, seed = struct.unpack("<QQQ", data[0:offset])
+
+    if length != len(data) - LEGACY_OFFSET:
+        # This can happen if the data was serialized on Windows, where the units
+        # were 32bit rather than 64bit.
+        unit = "L"
+        unit_size = LEGACY_WINDOWS_UNIT_SIZE
+        offset = LEGACY_WINDOWS_OFFSET
+        hcount, length, seed = struct.unpack("<LLL", data[0:offset])
+
+        # The length was the number of bytes in memory. But because of the
+        # platform size issue, the actual serialized bytes is half that.
+        assert length // 2 == len(data) - offset, "Length is invalid"
+
+    # This is the same in either case because the Windows code was written
+    # without being aware of the difference in size between Windows and
+    # non-Windows.
+    decode_len = length // LEGACY_UNIT_SIZE
+
+    bloom.hcount = hcount
+    bloom.length = length
+    bloom.seed = <uint32_t>seed
+
+    # This is tricky - to remove empty space we're going to map bytes into
+    # containers. On Windows or Linux, length is both the number of significant
+    # bits in the bitfield and the number of bytes when the bitfield was in
+    # memory in the old format. In our output, length will be the bitfield
+    # length in bits.
+    buflen = length // KEY_BITS
+    if length % KEY_BITS > 0:
+        buflen += 1
+    contents = struct.unpack(f"<{decode_len}{unit}", data[offset:])
+    assert buflen > 0, "Tried to allocate an empty buffer"
+    bloom.bitfield = <key_t*>mem.alloc(buflen, sizeof(key_t))
+
+    # Each item in contents provides one significant byte, so we'll copy it
```

Comment:
A couple of comments use _significant bits/byte_, this confused me, because the bits that were populated in the old storage are the least significant bits/byte? E.g.:

```
        hv = (keys[0] + (hiter * keys[1])) % bloom.length
        bloom.bitfield[hv // sizeof(key_t)] |= one << (hv % sizeof(key_t))
```

E.g. if `hv=18`:

```
one << (hv % sizeof(key_t))
1 << (18 % 8)
1 << 2
00000000 00000000 00000000 00000100
```

Reply:
Ah, that's correct - I should have used different phrasing, because I mean something like "non-zero / actually used bits" here. I'll revise that...